### PR TITLE
[CARBONDATA-2441][Datamap] Implement distribute interface for bloom datamap

### PR DESCRIPTION
**Pull request by xuchuanyin**
implement distribute interface for bloom datamap

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 `No`
 - [ ] Any backward compatibility impacted?
`No` 
 - [ ] Document update required?
`No`
 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
`Tests added`
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5658/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4498/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4735/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5690/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4530/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5703/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4543/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4768/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4781/

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomDataMapDistributable.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+package org.apache.carbondata.datamap.bloom;
+
+import org.apache.carbondata.common.annotations.InterfaceAudience;
+import org.apache.carbondata.core.datamap.DataMapDistributable;
+
+@InterfaceAudience.Internal
+class BloomDataMapDistributable extends DataMapDistributable {
+
+  // TODO: seems no one use this?
+  private String dataPath;
```
Comment (jackylk) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomDataMapDistributable.java]: Remove it if no one use it

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java [hunk 2]
```diff
@@ -154,12 +158,68 @@ public DataMapRefresher createRefresher(Segment segment, String shardName) throw
...
+   */
+  private CarbonFile[] getAllIndexDirs(String tablePath, String segmentId) {
+    List<CarbonFile> indexDirs = new ArrayList<>();
+    List<TableDataMap> dataMaps = new ArrayList<>();
+    try {
+      // there can be multiple bloom datamaps present on a table, so get all datamaps and form
+      // the path till the index file directories in all datamaps folders present in each segment
+      dataMaps = DataMapStoreManager.getInstance().getAllDataMap(getCarbonTable());
+    } catch (IOException ex) {
+      LOGGER.error("failed to get datamaps");
```
Comment (jackylk) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java]: should throw ex

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java [hunk 3]
```diff
@@ -154,12 +158,68 @@ public DataMapRefresher createRefresher(Segment segment, String shardName) throw
...
       throws IOException {
-    return null;
+    List<CoarseGrainDataMap> coarseGrainDataMaps = new ArrayList<>();
+    BloomCoarseGrainDataMap bloomCoarseGrainDataMap = new BloomCoarseGrainDataMap();
+    String indexPath = ((BloomDataMapDistributable) distributable).getIndexPath();
+    try {
+      bloomCoarseGrainDataMap.init(new DataMapModel(indexPath));
+    } catch (IOException e) {
+      LOGGER.error(e, "Failed to get bloom datamap");
+      return coarseGrainDataMaps;
```
Comment (jackylk) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java]: why not throw e?

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomDataMapDistributable.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+package org.apache.carbondata.datamap.bloom;
+
+import org.apache.carbondata.common.annotations.InterfaceAudience;
+import org.apache.carbondata.core.datamap.DataMapDistributable;
+
+@InterfaceAudience.Internal
+class BloomDataMapDistributable extends DataMapDistributable {
+  private String indexPath;
+
+  BloomDataMapDistributable(String dataPath, String indexPath) {
```
Comment (jackylk) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomDataMapDistributable.java]: dataPath is not used

Comment (jackylk) [reply datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomDataMapDistributable.java]: And can you describe what should be the indexPath? file path or folder path?

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java [hunk 5]
```diff
@@ -154,12 +158,65 @@ public DataMapRefresher createRefresher(Segment segment, String shardName) throw
...
+    BloomCoarseGrainDataMap bloomCoarseGrainDataMap = new BloomCoarseGrainDataMap();
+    String indexPath = ((BloomDataMapDistributable) distributable).getIndexPath();
+    bloomCoarseGrainDataMap.init(new DataMapModel(indexPath));
+    coarseGrainDataMaps.add(bloomCoarseGrainDataMap);
+    return coarseGrainDataMaps;
+  }
+
+  /**
+   * returns all the directories of lucene index files for query
+   * Note: copied from luceneDataMapFactory, will extract to a common interface
```
Comment (jackylk) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java]: Add a TODO

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5736/

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java (hunk 5 above)
Comment (xuchuanyin) [reply datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMapFactory.java]: Will do it after your PR #2255 is merged

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5744/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4807/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4586/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**